File: custom_adapter/TopOnAdapter/Objc/6.5.x/AlxTopOnBaseManager.m
//
//  AlxTopOnBaseManager.m
//  AlxAdsOCDemo
//

#import "AlxTopOnBaseManager.h"
#import <AnyThinkSDK/AnyThinkSDK.h>
#import <AlxAds/AlxAds.h>
#import "AlxTopOnMetaInfo.h"

static NSString *const TAG = @"AlxTopOnBaseManager";
static BOOL _isInitialized = NO;

@implementation AlxTopOnBaseManager

+ (void)setIsInitialized:(BOOL)isInitialized {
    _isInitialized = isInitialized;
}

+ (BOOL)isInitialized {
    return _isInitialized;
}

+ (NSString *)unitID {
    return @"slot_id";
}

+ (NSString *)initSDKWithServerInfo:(NSDictionary *)serverInfo {
    NSLog(@"%@: alx-sdk-version:%@", TAG, [AlxSdk getSDKVersion]);
    NSLog(@"%@: topon-sdk-version:%@", TAG, [[ATAPI sharedInstance] version]);
    NSLog(@"%@: topon-adapter-version:%@", TAG, [AlxTopOnMetaInfo ADAPTER_VERSION]);
    
    NSString *appid = serverInfo[@"appid"];
    NSString *sid = serverInfo[@"sid"];
    NSString *token = serverInfo[@"token"];
    NSString *debug = serverInfo[@"isdebug"];
    
    NSLog(@"%@: token=%@; appid=%@; sid=%@", TAG, token ?: @"", appid ?: @"", sid ?: @"");
    
    if (!appid || !sid || !token) {
        NSString *errorStr = @"initialize alx params: appid or sid or token is empty";
        NSLog(@"%@: error: %@", TAG, errorStr);
        return errorStr;
    }
    
    [AlxSdk initializeSDKWithToken:token sid:sid appId:appid];
    [self setIsInitialized:YES];
    [self sdkInfo];
    
    if (debug.length > 0) {
        if ([debug.lowercaseString isEqualToString:@"true"]) {
            [AlxSdk setDebug:YES];
        } else if ([debug.lowercaseString isEqualToString:@"false"]) {
            [AlxSdk setDebug:NO];
        }
    }
    
    // User Privacy - GDPR Consent Handling
    NSInteger gdprFlag = [[NSUserDefaults standardUserDefaults] integerForKey:@"IABTCF_gdprApplies"];
    NSString *gdprConsent = [[NSUserDefaults standardUserDefaults] stringForKey:@"IABTCF_TCString"];
    
    if (gdprFlag == 1) {
        [AlxSdk setGDPRConsent:YES];
    } else {
        [AlxSdk setGDPRConsent:NO];
    }
    [AlxSdk setGDPRConsentMessage:gdprConsent ?: @""];
    
    return nil;
}

+ (void)sdkInfo {
    NSDictionary *data = @{
        @"sdk_name": @"TopOn",
        @"sdk_version": [[ATAPI sharedInstance] version],
        @"adapter_version": [AlxTopOnMetaInfo ADAPTER_VERSION]
    };
    [AlxSdk addExtraParametersWithKey:@"alx_adapter" value:data];
}

+ (NSError *)errorWithCode:(NSInteger)code message:(NSString *)message {
    return [NSError errorWithDomain:@"AlxTopOnAdapter"
                               code:code
                           userInfo:@{NSLocalizedDescriptionKey: message}];
}

@end

File: Examples/SwiftDemo/RixSDKSPMDemo/AlxAdsDemo/CustomAdapter/TopOnAdapter/AlxTopOnBaseManager.swift
//
//  AlxTopOnBaseManager.swift
//

import Foundation
import AnyThinkSDK
import AlxAds

@objc(AlxTopOnBaseManager)
public class AlxTopOnBaseManager:NSObject {
    
    private static let TAG = "AlxTopOnBaseManager"
    
    public static var isInitialized:Bool = false
    
    public static let unitID = "slot_id"
    
    @discardableResult
    public static func initSDK(serverInfo: [AnyHashable: Any]) -> String?{
        NSLog("%@: alx-sdk-version:%@",AlxTopOnBaseManager.TAG,AlxSdk.getSDKVersion())
        NSLog("%@: topon-sdk-version:%@",AlxTopOnBaseManager.TAG,ATAPI.sharedInstance().version())
        NSLog("%@: topon-adapter-version:%@",AlxTopOnBaseManager.TAG,AlxTopOnMetaInf.ADAPTER_VERSION)
        
        let appid:String? = serverInfo["appid"] as? String
        let sid:String? = serverInfo["sid"] as? String
        let token:String? = serverInfo["token"] as? String
        let debug:String? = serverInfo["isdebug"] as? String
        
        NSLog("%@: token=%@; appid=%@; sid=%@",AlxTopOnBaseManager.TAG,token ?? "",appid ?? "",sid ?? "")
        guard let appid=appid,let sid=sid,let token=token else{
            let errorStr="initialize alx params: appid or sid or token is empty"
            NSLog("%@: error: %@",AlxTopOnBaseManager.TAG,errorStr)
            return errorStr
        }
        
        AlxSdk.initializeSDK(token: token, sid: sid, appId: appid)
        AlxTopOnBaseManager.isInitialized = true
        self.sdkInfo()
        
        if let debug,!debug.isEmpty {
            if debug.lowercased() == "true" {
                AlxSdk.setDebug(true)
            }else if debug.lowercased() == "false" {
                AlxSdk.setDebug(false)
            }
        }
        
        
        // User Privacy
        // MARK: - GDPR Consent Handling
        let gdprFlag = UserDefaults.standard.integer(forKey: "IABTCF_gdprApplies")
        let gdprConsent = UserDefaults.standard.string(forKey: "IABTCF_TCString")
        if gdprFlag == 1{
            AlxSdk.setGDPRConsent(true)
        }else  {
            AlxSdk.setGDPRConsent(false)
        }
        AlxSdk.setGDPRConsentMessage(gdprConsent ?? "")
        
        return nil
    }
    
    public static func sdkInfo(){
        var data:[String:String] = [:]
        data["sdk_name"] = "TopOn"
        data["sdk_version"] = ATAPI.sharedInstance().version()
        data["adapter_version"] = AlxTopOnMetaInf.ADAPTER_VERSION
        AlxSdk.addExtraParameters(key:"alx_adapter",value:data)
    }
    
    public static func error(code:Int,msg:String) -> NSError{
        return NSError(domain: "AlxTopOnAdapter", code: code, userInfo: [NSLocalizedDescriptionKey : msg])
    }
    
}
